### PR TITLE
initial commit of the yurttunnel

### PR DESCRIPTION
**Pull request by charleszheng44**
1. add the main body of the yurttunel-server and the yurttunnel-agent
2. add implementations of tool functions

### CODE REVIEW

File: pkg/yurttunnel/agent/cmd.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+	o := &YurttunnelAgentOptions{
+		kubeConfig: "/etc/kubernetes/kubelet.conf",
+		caFile:     "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt",
+	}
+	return o
+}
+
+// validate validates the YurttunnelServerOptions
+func (o *YurttunnelAgentOptions) validate() error {
+	if o.kubeConfig == "" {
```
Comment (rambohe-ch) [inline pkg/yurttunnel/agent/cmd.go:85]: the condition is o.kubeConfig != "" ?

File: pkg/yurttunnel/agent/cmd.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+			return fmt.Errorf("%s doesn't exist", o.kubeConfig)
+		}
+		if o.caFile == "" {
+			return errors.New("either '--kube-config' or " +
+				"'--ca-file' need to be set")
+		}
+		if _, err := os.Stat(o.caFile); os.IsNotExist(err) {
+			return fmt.Errorf("%s doesn't exist", o.caFile)
+		}
+		if o.serverAddr == "" {
```
Comment (rambohe-ch) [inline pkg/yurttunnel/agent/cmd.go:96]: serverAddr is the address of yurt tunnel server, and kubeConfig is rest config for accessing kube-apiserver, both these two parameters should be set ?

File: pkg/yurttunnel/agent/cmd.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+		if err != nil {
+			return err
+		}
+	} else {
+		serverAddr = o.serverAddr
+	}
+	klog.Infof("tunnel server address: %s", serverAddr)
+
+	// 2. create a certificate manager
+	config, err := rest.InClusterConfig()
```
Comment (rambohe-ch) [inline pkg/yurttunnel/agent/cmd.go:128]: InClusterConfig can not be used by pods that run on edge node for pods can not access kube-apiserver by KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT setting

File: pkg/yurttunnel/agent/cmd.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+func (o *YurttunnelAgentOptions) validate() error {
+	if o.kubeConfig == "" {
+		if _, err := os.Stat(o.kubeConfig); os.IsNotExist(err) {
+			return fmt.Errorf("%s doesn't exist", o.kubeConfig)
+		}
+		if o.caFile == "" {
+			return errors.New("either '--kube-config' or " +
+				"'--ca-file' need to be set")
+		}
+		if _, err := os.Stat(o.caFile); os.IsNotExist(err) {
```
Comment (rambohe-ch) [inline pkg/yurttunnel/agent/cmd.go:93]: when ca file is not exist, would you get is from kube-apiserver?

File: pkg/yurttunnel/agent/cmd.go (hunk 1 above)
Comment (charleszheng44) [reply pkg/yurttunnel/agent/cmd.go:85]: will fix

File: pkg/yurttunnel/agent/cmd.go (hunk 2 above)
Comment (charleszheng44) [reply pkg/yurttunnel/agent/cmd.go:96]: if serverAddr is not set, we will get it through the function GetServerAddr. if kubeconfig not set, we will create a incluster config? So if neither of them are set, the agent can still work?

File: pkg/yurttunnel/agent/cmd.go (hunk 3 above)
Comment (charleszheng44) [reply pkg/yurttunnel/agent/cmd.go:128]: will fix

File: pkg/yurttunnel/agent/cmd.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+			return nil
+		},
+	}
+
+	flags := cmd.Flags()
+	flags.StringVar(&o.nodeName, "node-name", o.nodeName,
+		"The name of the edge node.")
+	flags.StringVar(&o.serverAddr, "server-addr", o.serverAddr,
+		"The address of yurttunnel-server")
+	flags.StringVar(&o.kubeConfig, "kube-config", o.kubeConfig,
```
Comment (rambohe-ch) [inline pkg/yurttunnel/agent/cmd.go:57]: maybe we need add a kubeServerAddr parameter for accessing kube-apiserver when kubeConfig is not set or kubeConfig not exist

File: pkg/yurttunnel/agent/cmd.go (hunk 4 above)
Comment (charleszheng44) [reply pkg/yurttunnel/agent/cmd.go:93]: You mean from kubeconfig?

File: pkg/yurttunnel/server/cmd.go [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+		"server will accept master requests at: %s",
+		o.serverAgentAddr, o.serverMasterAddr)
+}
+
+// run starts the yurttunel-server
+func (o *YurttunnelServerOptions) run(stopCh <-chan struct{}) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	// 1. create the master's clientset
+	clientSet, csErr := kubernetes.CreateClientSet(o.kubeConfig)
```
Comment (rambohe-ch) [inline pkg/yurttunnel/server/cmd.go:120]: kubeConfig may not exist, could we add another parameter as a backup?

Comment (charleszheng44) [reply pkg/yurttunnel/server/cmd.go]: `kubeconfig` will be a mandatory option.

File: pkg/yurttunnel/agent/cmd.go (hunk 5 above)
Comment (charleszheng44) [reply pkg/yurttunnel/agent/cmd.go:66]: if we want to generate the clientset (or rest.Config) based on the apiserver address, we will also need the certificate, which is generated by certmanager. However, certmanager can only be created if a clientset is available. There exist a mutual dependency here. May be we should make `kubeconfig` as a mandatory option?

File: pkg/yurttunnel/agent/cmd.go (hunk 2 above)
Comment (charleszheng44) [reply pkg/yurttunnel/agent/cmd.go]: if `serverAddr` is not set, we can get it by calling `GetServerAddr` function

File: pkg/yurttunnel/pki/pki.go [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+		return nil, fmt.Errorf("loading CA from %s fail: %s", caFile, err)
+	}
+	caPEM, err := ioutil.ReadFile(caFile)
+	if err != nil {
+		return nil, err
+	}
+
+	rootCertPool := x509.NewCertPool()
+	rootCertPool.AppendCertsFromPEM(caPEM)
+	return rootCertPool, nil
```
Comment (rambohe-ch) [inline pkg/yurttunnel/pki/pki.go:124]: would you use GenCertPoolUseCA func to refactor the code snippet?

Comment (charleszheng44) [reply pkg/yurttunnel/pki/pki.go]: fixed

**Comment by rambohe-ch** [comment]
lgtm

**Pull request merged**